Source/DafnyCore/DafnyFile.cs

```diff
@@ -16,13 +18,30 @@ public class DafnyFile {
   public string BaseName { get; private set; }
   public bool IsPreverified { get; set; }
   public bool IsPrecompiled { get; set; }
-  public bool IsPrerefined { get; private set; }
   public Func<TextReader> GetContent { get; set; }
-  public Uri Uri { get; }
+  public Uri Uri { get; private set; }
   [CanBeNull] public IToken Origin { get; }
 
+  private static readonly Dictionary<Uri, Uri> ExternallyVisibleEmbeddedFiles = new();
+
+  public static Uri ExposeInternalUri(string externalName, Uri internalUri) {
+    var externalUri = new Uri("dafny:" + externalName);
```

Comment:
Why not just use `$"dafny:{internalUri}"`? That would be simpler and leaves the door open for other schemes. AFAICT attaching a fake Dafny file name doesn't help the IDE but perhaps I'm missing something.

Reply:
Because we expose the externalUri and the internalUri contains implementation details. Is there are a downside to having this mapping, except for it being slightly more work?

Reply:
No real objection to the complexity/work, but a little worried about exposing "DafnyStandardLibraries.dfy" when that string/file name isn't actually meaningful. Eventually I expect the IDE will be able to actually open the doo file, and we shouldn't hide the fact that the source is coming from a doo file (especially to be clear it's read-only)

Reply:
Actually I added the dot dfy at the end so the IDE can recognize it's Dafny code and provide ide tooling on it.

I think from the user perspective, opening a dotdfy file to view the sources inside a dotdoo file, seems good

Reply:
Okay, I can live with it :)

Reply:
We can still change the part after the `dafny:` in a follow-up PR, so let's not get hung up on this

I think, as part of adding `.doo` files, we should have added IDE navigation for .doo files as well, and then we would have already solved this. I think it would help with correctly designing `.doo` files as well. For example, we could have considered not making it dotdoo a compressed file and instead having it be a single text file, so the IDE can open it directly, without the need for a `dafny doo source` command. Not saying we should have gone for that option, but at least we would have had a reason to consider it.

Thinking about dotdoo support in the IDE more, and given the way dotdoo files are currently structured, the `dafny doo source` command should support both on-disk files and files internal to the CLI, so it does need to take a URI, and then the `dafny:` uris exposed to the IDE should probably contain a URI after the `dafny:`. The IDE will chop off the `dafny:` part and pass the rest to `dafny doo source`

Reply:
> For example, we could have considered not making dotdoo a compressed file and instead having it be a single text file, so the IDE can open it directly, without the need for a dafny doo/library source command.

I definitely intended to iterate on the file format as we added features, and hence why the manifest has a `doo_file_version` property. :) But I don't think I'd want to make it a single text file, since then the manifest information would have to be embedded in comments or something like that, which would make it a bit tougher to read robustly and harder to hide that implementation detail from users.

> Thinking about dotdoo support in the IDE more, and given the way dotdoo files are currently structured, the dafny doo/library source command should support both on-disk files and files internal to the CLI, so it does need to take a URI, and then the dafny: uris exposed to the IDE should probably contain a URI after the dafny:. The IDE will chop off the dafny: part and pass the rest to dafny doo source

Yup, this is one of the reasons I thought `$"dafny:{internalUri}"` would be smart, but you expressed it better than I managed at the time. :)